Support Proxy Nodes and policy for external references

ftomassetti opened the pull request:
In this PR we cover a problem similar to what we covered in #132 

In #132 we looked at what happens when we retrieve a partial tree (so we have references to a parent that we have not retrieved). In that case we introduced the possibility of picking a policy to specify what to do: should we throw an error? Or set the un-retrieved parent to null?

We now introduce a third possibility: generating proxy nodes.

With this PR we also introduce the possibility to set this policy not only to deal with parents that are unavailable, but also to deal with references to nodes that are not available (which is typically the case when referring to nodes outside the tree considered). When retrieving a tree we want references which are local to the tree to be resolved correctly, setting the "real node" as the referred value. For references to nodes outside the tree consider we can again specify a policy to deal with that. The three options are the same: throw error, set them to null, or use proxy nodes.

enikao commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -680,11 +806,22 @@ private void populateNodeReferences(
-                                  + entry.getReference()
-                                  + " for feature "
-                                  + serializedReferenceValue.getMetaPointer());
+                          if (unavailableReferenceTargetPolicy
```
Use `switch` with enum

ftomassetti replied:
Done

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/model/impl/ProxyNode.java:
```diff
@@ -0,0 +1,116 @@
+
+  private String id;
+
+  public ProxyNode(String id) {
```
Annotation added

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/model/impl/ProxyNode.java:
```diff
@@ -0,0 +1,116 @@
+
+  @Nullable
+  @Override
+  public String getID() {
```
Annotation added

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -423,56 +446,137 @@ private void validateSerializationBlock(SerializedChunk serializationBlock) {
+      // effectively treated as roots and their parent will be set to an instance of a ProxyNode,
+      // as we cannot retrieve them or set them (until we decide to provide some sort of
+      // NodeResolver)
+      Set<String> knownIDs =
```
Right, we have already Guava among dependencies so I have updated it and then used it to simplify this code

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -529,8 +634,25 @@ private List<ClassifierInstance<?>> deserializeClassifierInstances(
+                  sortingResult);
               ClassifierInstance<?> classifierInstance = serializedToInstanceMap.get(node);
+              if (unavailableParentPolicy == UnavailableNodePolicy.PROXY_NODES) {
+                // For real parents, the parent is not set directly, but it is set indirectly
```
I am not sure I have a reasonable solution for that. We could set the children we know of in the proxy parent node, but then the proxy would have only a _partial_ list of children (just the ones that we are aware of, as we do not retrieve the data of the proxy node). So we could use the proxy nodes to add children in it only for the purpose of it triggering `setParent` on those children. I do not think this would make the situation better and I cannot think of other options

enikao replied:
I agree, doesn't sound better.

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -423,56 +446,137 @@ private void validateSerializationBlock(SerializedChunk serializationBlock) {
+    sortingResult.putNodesWithNullIDsInFront();
 
-    if (unknownParentPolicy == UnknownParentPolicy.NULL_REFERENCES) {
+    if (unavailableParentPolicy == UnavailableNodePolicy.NULL_REFERENCES) {
```
Done

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -617,9 +740,11 @@ private ClassifierInstance<?> instantiateFromSerialized(
       SerializedClassifierInstance serializedClassifierInstance,
       ClassifierInstance<?> node,
-      ClassifierInstanceResolver classifierInstanceResolver) {
+      ClassifierInstanceResolver classifierInstanceResolver,
```
I have extracted a supporting class that I have called NodePopulator, for lack of a better name

ftomassetti commented on core/src/main/java/io/lionweb/lioncore/java/serialization/JsonSerialization.java:
```diff
@@ -423,56 +446,137 @@ private void validateSerializationBlock(SerializedChunk serializationBlock) {
+   * while sorting this particular set of nodes, as the proxies created in this process will need to
+   * be returned together with the sorted list of unserialized nodes.
+   */
+  private class SortingResult {
```
I have renamed it to DeserializationStatus and I have extracted it

ftomassetti commented:
@enikao thank you for the review: it helped improving the code significantly. I think I have addressed all of your comments with the exception of the one about getting rid of the asymmetry. For that one I have not a valid solution

enikao commented on core/src/main/java/io/lionweb/lioncore/java/serialization/DeserializationStatus.java:
```diff
@@ -0,0 +1,74 @@
+ * list of unserialized nodes.
+ */
+class DeserializationStatus {
+  List<SerializedClassifierInstance> sortedList;
```
I think all of these fields can be `final`?

ftomassetti replied:
Making all these fields final in follow up commit

ftomassetti commented:
Thank you, I interpret the thumbs up as approval for this PR and I will go on and merge it. In case there are other issues I missed please let me know

enikao commented on core/src/main/java/io/lionweb/lioncore/java/serialization/NodePopulator.java:
```diff
@@ -0,0 +1,116 @@
+ * deserializing it.
+ */
+class NodePopulator {
+  private JsonSerialization jsonSerialization;
```
`final`?

ftomassetti replied:
Making all these fields final in follow up commit

enikao commented on core/src/main/java/io/lionweb/lioncore/java/serialization/NodePopulator.java:
```diff
@@ -0,0 +1,116 @@
+                                      + entry.getReference()
+                                      + " for feature "
+                                      + serializedReferenceValue.getMetaPointer());
+                            default:
```
We don't need this, and it's actually harmful: If we would ever add a new literal to the enum, the compiler would tell us.
